### PR TITLE
Upgrade test VM to Centos 7.1 and add more disk space

### LINKED ISSUE #419: Test VM: VirtualBox guest additions fail to build (missing kernel headers)
The test VM is failing to come up on `vagrant up` as it looks like the kernel headers are now no longer on the CentOS 6.6 image (and the same issue occurs when trying to use CentOS 7).

Also, the headers for the CentOS 6.6 image's kernel version on the VM don't seem to be available to the default yum repositories. 

The work-around is to use a version of VirtualBox that matches the guest additions that are already on the CentOS image (4.3.26 works) or to manually install the headers. E.g. with

` yum install ftp://mirror.switch.ch/pool/4/mirror/scientificlinux/6.6/x86_64/os/Packages/kernel-devel-2.6.32-504.el6.x86_64.rpm`

and then `vagrant reload`.

I'm not sure this is possible to fix in the DbFit VM provisioning. I.e. is it possible to inject this `yum` command before the guest additions get built?

I'm not sure this is desirable either as it's more stuff to have to maintain.

Any views?

### PR DESCRIPTION
**Pull request by MMatten**
* Upgrade CentOS on the test VM to 7.1.
* Adds 10GB of available disk space to the existing file system.
* Resolves #419.

### CODE REVIEW

File: test_vm/Vagrantfile [hunk 1]
```diff
@@ -34,7 +46,7 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
         "password" => {
           "postgres" => "postgres"
         },
-        "version" => "9.3",
+        "version" => "9.4",
```
Comment (MMatten) [inline test_vm/Vagrantfile]: I've got one the Postgres upgrade related changes from the `test-vm-update-postgresql-to-9.4` branch.

File: test_vm/bootstrap.sh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Create new physical volume.
+pvcreate /dev/sdb
+# Add the physical volume to the existing volume group.
+vgextend centos /dev/sdb
+# Add the physical volume to the logical volume.
+lvextend /dev/centos/root /dev/sdb
+# Resize the file system to use the new space available in the logical volume..
+xfs_growfs /dev/centos/root
```
Comment (MMatten) [inline test_vm/bootstrap.sh]: Hmmm...I still need to add some more error trapping here (test the exit status of all of the commands).

Comment (javornikolov) [inline test_vm/bootstrap.sh]: Is the file system indeed xfs (I think the default is ext3 or ext4)? Can we check the output of `mount` and `cat /etc/fstab`?

Comment (javornikolov) [inline test_vm/bootstrap.sh]: I can see there is a chef cookbook for lvm management: https://supermarket.chef.io/cookbooks/lvm

Comment (javornikolov) [inline test_vm/bootstrap.sh]: And also for fs resize: https://supermarket.chef.io/cookbooks/filesystem_resize/versions/0.1.0

File: test_vm/Vagrantfile [hunk 3]
```diff
@@ -5,21 +5,33 @@ require_relative 'vagrant_config_helper'
 
 VAGRANTFILE_API_VERSION = "2"
 
+file_to_disk = "./disk1.vdi"
```
Comment (javornikolov) [inline test_vm/Vagrantfile]: Is this path relative to dbfit project folder or to the VirtualBox/vagrant disks?

File: test_vm/Vagrantfile (hunk 1 above)
Comment (javornikolov) [inline test_vm/Vagrantfile]: I think postgresql upgrade would also need the customized cookbook reflected in Cheffile, Cheffile.lock? Maybe we can get these through a separate PR.

Comment (javornikolov) [inline test_vm/Vagrantfile]: > I think postgresql upgrade would also need the customized cookbook reflected in Cheffile, Cheffile.lock? Maybe we can get these through a separate PR.

Done in #422. I'll merge so that we can rebase and concentrate here on the disk resizing issue.

File: test_vm/bootstrap.sh (hunk 2 above)
Comment (MMatten) [inline test_vm/bootstrap.sh]: > Is the file system indeed xfs (I think the default is ext3 or ext4)? Can we check the output of `mount` and `cat /etc/fstab`?

Yes, definitely. `resize2fs` failed. The disk space created by the commands is usable as I was able to install Oracle and DB2 and run the integration tests.

I'll have a go with the Chef option now though.

File: test_vm/Vagrantfile (hunk 3 above)
Comment (MMatten) [inline test_vm/Vagrantfile]: > Is this path relative to dbfit project folder or to the VirtualBox/vagrant disks?

This is relative to the `test_vm` directory.

Comment (javornikolov) [inline test_vm/Vagrantfile]: May we have this configurable? E.g. - via `vagrant_config_custom.rb`

File: test_vm/bootstrap.sh (hunk 2 above)
Comment (MMatten) [inline test_vm/bootstrap.sh]: I seem to be hitting the caveat mentioned at https://github.com/opscode-cookbooks/lvm and I'm in unfamiliar territory again.

Do you think it's worth following that trail or just for the the shell method?

Comment (javornikolov) [inline test_vm/bootstrap.sh]: Hmm. Let's initially go on with the shell method. We can switch to the lvm cookbook later after we fix the lvm attributes issue.

Comment (MMatten) [inline test_vm/bootstrap.sh]: I've just created the following PR to support the LVM version on the CentoOS box image: https://github.com/gregsymons/di-ruby-lvm-attrib/pull/29

I'll see if this helps...

Comment (MMatten) [inline test_vm/bootstrap.sh]: Ok, the `di-ruby-lvm-attrib` PR wasn't actually required.

A new one at https://github.com/opscode-cookbooks/lvm/pull/67 is pending.

**Comment by MMatten** [comment]
Ok, back from a break...

I've got the logical volume / file system extending to use additional space via the Chef `LVM` cookbook now.

But I'm hitting another Postgres issue: -



==> default:   ================================================================================
==> default:   Recipe Compile Error in /tmp/vagrant-chef/fef29433d03cac37dc603db1ce2d9b9d/cookbooks/dbfit_test/recipes/default.rb
==> default:   ================================================================================
==> default:
==> default:   KeyError
==> default:   --------
==> default:   key not found: "centos"
==> default:
==> default:   Cookbook Trace:
==> default:   ---------------
==> default:     /tmp/vagrant-chef/fef29433d03cac37dc603db1ce2d9b9d/cookbooks/postgresql/libraries/default.rb:360:in `pgdgrepo_rpm_i
nfo'
==> default:     /tmp/vagrant-chef/fef29433d03cac37dc603db1ce2d9b9d/cookbooks/postgresql/recipes/yum_pgdg_postgresql.rb:31:in `from_
file'
==> default:     /tmp/vagrant-chef/fef29433d03cac37dc603db1ce2d9b9d/cookbooks/postgresql/recipes/client.rb:27:in `from_file'
==> default:     /tmp/vagrant-chef/fef29433d03cac37dc603db1ce2d9b9d/cookbooks/postgresql/recipes/server.rb:20:in `from_file'
==> default:     /tmp/vagrant-chef/fef29433d03cac37dc603db1ce2d9b9d/cookbooks/dbfit_test/recipes/postgres.rb:1:in `from_file'
==> default:     /tmp/vagrant-chef/fef29433d03cac37dc603db1ce2d9b9d/cookbooks/dbfit_test/recipes/default.rb:4:in `from_file'
==> default:
==> default:   Relevant File Content:
==> default:   ----------------------
==> default:   /tmp/vagrant-chef/fef29433d03cac37dc603db1ce2d9b9d/cookbooks/postgresql/libraries/default.rb:
==> default:
==> default:   353:  # since PGDG supports several versions of PostgreSQL, platforms, platform versions
==> default:   354:  # and architectures.
==> default:   355:  # Links to RPMs for installation are in an attribute so that new versions/platforms
==> default:   356:  # can be more easily added. (See attributes/default.rb)
==> default:   357:  def pgdgrepo_rpm_info
==> default:   358:    repo_rpm_url = node['postgresql']['pgdg']['repo_rpm_url'].
==> default:   359:      fetch(node['postgresql']['version']).            # e.g., fetch for "9.1"
==> default:   360>>     fetch(node['platform']).                         # e.g., fetch for "centos"
==> default:   361:      fetch(node['platform_version'].to_f.to_i.to_s).  # e.g., fetch for "5" (truncated "5.7")
==> default:   362:      fetch(node['kernel']['machine'])                 # e.g., fetch for "i386" or "x86_64"
==> default:   363:
==> default:   364:    # Extract the filename portion from the URL for the PGDG repository RPM.
==> default:   365:    # E.g., repo_rpm_filename = "pgdg-centos92-9.2-6.noarch.rpm"
==> default:   366:    repo_rpm_filename = File.basename(repo_rpm_url)
==> default:   367:
==> default:   368:    # Extract the package name from the URL for the PGDG repository RPM.
==> default:   369:    # E.g., repo_rpm_package = "pgdg-centos92"
==> default:
==> default:
==> default:   Running handlers:
==> default: [2015-09-03T14:58:23+00:00] ERROR: Running exception handlers
==> default:   Running handlers complete
==> default: [2015-09-03T14:58:23+00:00] ERROR: Exception handlers complete
==> default:   Chef Client failed. 1 resources updated in 12.316700784 seconds
==> default: [2015-09-03T14:58:23+00:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
==> default: [2015-09-03T14:58:23+00:00] ERROR: key not found: "centos"
==> default: [2015-09-03T14:58:23+00:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit c
ode 1)
Chef never successfully completed! Any errors should be visible in the
output above. Please fix your recipes so that they properly complete.


@javornikolov, do you recognise this one?

**Comment by MMatten** [comment]
@javornikolov - ok, let me try again using the `https://github.com/dbfit/postgresql` fork. I was still referencing the source version, even after rebasing (I hadn't let Notepad++ reload the file before saving an experimental change).

**Comment by MMatten** [comment]
All appears to be working well now.

**Comment by MMatten** [comment]
Oh, one more change to make: remove PAM install from DB2 shell script installer...

**Comment by MMatten** [comment]
All tested and happy with this now.

File: test_vm/Vagrantfile [hunk 4]
```diff
@@ -5,16 +5,23 @@ require_relative 'vagrant_config_helper'
 
 VAGRANTFILE_API_VERSION = "2"
 
+file_to_disk = CONFIG['additional_disk1_file']
+
 Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
   config.vm.provider "virtualbox" do |vb|
     vb.customize ["modifyvm", :id, "--memory", CONFIG['memory']]
+    # Add another disk.
+    if ARGV[0] == "up" && ! File.exist?(file_to_disk)
```
Comment (javornikolov) [inline test_vm/Vagrantfile]: Let's remove the space after `!` (https://github.com/bbatsov/ruby-style-guide)

File: test_vm/Cheffile [hunk 5]
```diff
@@ -8,3 +8,4 @@ cookbook 'dbfit_test', :path => "vendor/cookbooks/dbfit_test"
 cookbook 'database', :git => "https://github.com/dbfit/database"
 cookbook 'postgresql', :git => "https://github.com/dbfit/postgresql", :ref => "develop"
 cookbook 'swap', :git => "https://github.com/dbfit/swap"
+cookbook 'lvm', :git => "http://github.com/opscode-cookbooks/lvm"
```
Comment (javornikolov) [inline test_vm/Cheffile]: Do we need to explicitly configure that, Is the default lvm cookbook failing?

File: test_vm/Cheffile [hunk 6]
```diff
@@ -8,3 +8,4 @@ cookbook 'dbfit_test', :path => "vendor/cookbooks/dbfit_test"
 cookbook 'database', :git => "https://github.com/dbfit/database"
```
Comment (javornikolov) [inline test_vm/Cheffile]: Maybe a bit off-topic here, but some PRs have been recently merged into the upstream `database` git repo. Maybe it's already good enough and no need to customize the cookbook location.

**Comment by javornikolov** [comment]
> All tested and happy with this now.

Sounds great!
Is the added disk space allocated in advance (all the 10G), or will dynamically grow when it's used?

File: test_vm/Cheffile (hunk 5 above)
Comment (MMatten) [inline test_vm/Cheffile]: Ah, I see. I didn't realise that didn't need to be specified. I'll remove that.

File: test_vm/Cheffile (hunk 6 above)
Comment (MMatten) [inline test_vm/Cheffile]: I'll see if I can remove this now.

**Comment by MMatten** [comment]
> Is the added disk space allocated in advance (all the 10G), or will dynamically grow when it's used?

Yes, it grows dynamically. It's about 1.3GB on my machine. One observation is that the provisioning is a bit slower; I'm guessing because of the disk needs to grow during setup but I'd expect performance to remain constant after that.

File: test_vm/Cheffile (hunk 6 above)
Comment (MMatten) [inline test_vm/Cheffile]: Ok, everything updated and tested apart from this now. I'll have to do that tomorrow.

File: test_vm/vendor/cookbooks/dbfit_test/recipes/diskspace.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
+chef_gem 'di-ruby-lvm'
+
+lvm_physical_volume '/dev/sdb' do
+  action :create
+end
+
+lvm_volume_group 'centos' do
+  action :extend
+  physical_volumes ['/dev/sda2', '/dev/sdb']
```
Comment (javornikolov) [inline test_vm/vendor/cookbooks/dbfit_test/recipes/diskspace.rb]: What is `/dev/sda2`, should we have it here? (Above we create only `/dev/sdb`).

Comment (MMatten) [inline test_vm/vendor/cookbooks/dbfit_test/recipes/diskspace.rb]: `/dev/sda2' is the existing LVM physical volume that comes with the CentOS 7.1. box.

I'm not sure if I need to specify all of the volumes, including existing volumes, in the volume group or to only specify the new ones. I'll try removing `/dev/sda2'.

**Comment by MMatten** [comment]
> Maybe a bit off-topic here, but some PRs have been recently merged into the upstream database git repo. Maybe it's already good enough and no need to customize the cookbook location.

Good news. The `database` cookbook seems to be working for us now to no need for our fork.

**Comment by MMatten** [comment]
Ok, updates applied. Can you spot anything else or ready to merge?

**Comment by javornikolov** [comment]
Looks good :+1:

**Comment by javornikolov** [comment]
Just 2nd commit includes Cheffile only while the related change of Cheffile.lock is in the 1st commit.

**Comment by MMatten** [comment]
> Just 2nd commit includes Cheffile only while the related change of Cheffile.lock is in the 1st commit.

Ah, yes. Is there value in seeing if this can be a separate commit, or even PR, or can we just squash it into the main PR commit?

**Comment by javornikolov** [comment]
> Ah, yes. Is there value in seeing if this can be a separate commit, or even PR, or can we just squash it into the main PR commit?

Let's squash it all in one commit.

**Comment by MMatten** [comment]
Squashed. 

Are you going to delete our `database` cookbook fork repo?

**Comment by javornikolov** [comment]
> Are you going to delete our database cookbook fork repo?

Maybe yes, since we don't need it anymore.

File: test_vm/vendor/cookbooks/dbfit_test/recipes/diskspace.rb [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+lvm_volume_group 'centos' do
+  action :extend
+  physical_volumes ['/dev/sdb']
+end
+
+lvm_logical_volume 'root' do
+  group 'centos'
+  action :resize
+  size '9G'
+  filesystem 'ext2'
```
Comment (javornikolov) [inline test_vm/vendor/cookbooks/dbfit_test/recipes/diskspace.rb]: Is it really ext2? I would rather expect ext3 or ext4. Is that parameter needed for resize?

Comment (MMatten) [inline test_vm/vendor/cookbooks/dbfit_test/recipes/diskspace.rb]: Actually, it looks like it's xfs. I'll see what else I can remove (possibly the mount point too).

**Comment by MMatten** [comment]
The mount point and file system type weren't required hence removed.

File: test_vm/vendor/cookbooks/dbfit_test/recipes/diskspace.rb [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+lvm_volume_group 'centos' do
+  action :extend
+  physical_volumes ['/dev/sdb']
+end
+
+lvm_logical_volume 'root' do
+  group 'centos'
+  action :resize
+  size '9G'
+  take_up_free_space true
```
Comment (javornikolov) [inline test_vm/vendor/cookbooks/dbfit_test/recipes/diskspace.rb]: I suspect we should have only one of the two size specifications: `size '9G'` and `take_up_free_space true`?

Comment (MMatten) [inline test_vm/vendor/cookbooks/dbfit_test/recipes/diskspace.rb]: Without the size we get an error: -

> ==> default: [2015-09-13T18:36:12+00:00] ERROR: lvm_logical_volume[root] (dbfit_test::diskspace line 12) had an error: Chef::Exceptions::ValidationFailed: Required argument size is missing!

I was expecting that we'd only need to specify `take_up_free_space true` to use all of the new space.

If I don't specify `take_up_free_space true` then the file system doesn't report the extra space being available and the provisioning runs out of disk space.

Comment (javornikolov) [inline test_vm/vendor/cookbooks/dbfit_test/recipes/diskspace.rb]: OK, I see.
